File: RemoveWWW/ApplicationHttpModule.cs
using System;
using System.Text.RegularExpressions;
using System.Web;

namespace RemoveWWW
	{

	public class ApplicationHttpModule : IHttpModule
	{
		
		public static string ModuleName
		{
			get
			{
				return "ApplicationHttpModule";
			}
		}
		// In the Init function, register for HttpApplication
		// events by adding your handlers.
		public void Init(HttpApplication application)
		{
			application.BeginRequest += Application_BeginRequest;
		}
		private void Application_BeginRequest(object source, EventArgs e)
		{
			// Create HttpApplication and HttpContext objects to access
			// request and response properties.
			HttpApplication application = (HttpApplication) source;
			HttpContext context = application.Context;
			string url = context.Request.Url.ToString();
			if (url.Contains("://www."))
			{
				RemoveWww(context);
			}
		}
		
		private static readonly Regex _Regex = new Regex("(http|https)://www\\.", (System.Text.RegularExpressions.RegexOptions) (RegexOptions.IgnoreCase | RegexOptions.Compiled));
		/// <summary>
		/// Removes the www subdomain from the request and redirects.
		/// </summary>
		private static void RemoveWww(HttpContext context)
		{
			string url = context.Request.Url.ToString();
			
			if (context.Request.RawUrl.Contains("404.aspx?404;"))
			{
				url = System.Convert.ToString(context.Request.RawUrl.Replace("/wpm/404.aspx?404;", string.Empty));
			}
			else
			{
				url = context.Request.Url.ToString();
			}
			if (_Regex.IsMatch(url))
			{
				url = _Regex.Replace(url, "$1://");
				PermanentRedirect(url, context);
			}
		}
		
		/// <summary>
		/// Sends permanent redirection headers (301)
		/// </summary>
		private static void PermanentRedirect(string url, HttpContext context)
		{
			if (url.EndsWith("default.aspx", StringComparison.OrdinalIgnoreCase))
			{
				url = System.Convert.ToString((url.ToLower()).Replace("default.aspx", string.Empty));
			}
			context.Response.Clear();
			context.Response.StatusCode = 301;
			context.Response.AppendHeader("location", url);
			context.Response.End();
		}
		public void Dispose()
		{
		}
	}
	
}

File: archive/WebProjectMechanics/Utility/HttpModule/ApplicationHttpModule.vb
Imports System.Text.RegularExpressions
Public Class ApplicationHttpModule
    Implements IHttpModule
    Private bRedirect As Boolean = False
    Public Shared ReadOnly Property ModuleName() As String
        Get
            Return "ApplicationHttpModule"
        End Get
    End Property
    ' In the Init function, register for HttpApplication 
    ' events by adding your handlers. 
    Public Sub Init(ByVal application As HttpApplication) _
            Implements IHttpModule.Init
        AddHandler application.BeginRequest, _
            AddressOf Application_BeginRequest
    End Sub
    Private Sub Application_BeginRequest(ByVal source As Object, _
             ByVal e As EventArgs)
        ' Create HttpApplication and HttpContext objects to access 
        ' request and response properties. 
        Dim application As HttpApplication = DirectCast(source, HttpApplication)
        Dim context As HttpContext = application.Context
        Dim url As String = context.Request.Url.ToString()
        If url.Contains("://www.") AndAlso wpm_SiteConfig.RemoveWWW Then
            RemoveWww(context)
            bRedirect = True
        End If
        If Not url.Contains("://www.") AndAlso Not wpm_SiteConfig.RemoveWWW Then
            AddWww(context)
            bRedirect = True
        End If
    End Sub

    ''' <summary>
    ''' Adds the www subdomain to the request and redirects.
    ''' </summary>
    Private Shared Sub AddWww(ByVal context As HttpContext)
        Dim url As String
        If InStr(context.Request.RawUrl(), "404.aspx?404;") > 0 Then
            url = Replace(context.Request.RawUrl(), wpm_SiteConfig.ApplicationHome & "404.aspx?404;", String.Empty)
        Else
            url = context.Request.Url.ToString()
        End If
        url = url.Replace("://", "://www.")
        PermanentRedirect(url, context)
    End Sub
    Private Shared ReadOnly _Regex As New Regex("(http|https)://www\.", RegexOptions.IgnoreCase Or RegexOptions.Compiled)
    ''' <summary>
    ''' Removes the www subdomain from the request and redirects.
    ''' </summary>
    Private Shared Sub RemoveWww(ByVal context As HttpContext)
        Dim url As String
        If InStr(context.Request.RawUrl(), "404.aspx?404;") > 0 Then
            url = Replace(context.Request.RawUrl(), wpm_SiteConfig.ApplicationHome & "404.aspx?404;", String.Empty)
        Else
            url = context.Request.Url.ToString()
        End If
        If _Regex.IsMatch(url) Then
            url = _Regex.Replace(url, "$1://")
            PermanentRedirect(url, context)
        End If
    End Sub

    ''' <summary>
    ''' Sends permanent redirection headers (301)
    ''' </summary>
    Private Shared Sub PermanentRedirect(ByVal url As String, ByVal context As HttpContext)
        If (url.EndsWith("default.aspx", StringComparison.OrdinalIgnoreCase)) Then
            url = Replace((url.ToLower), "default.aspx", String.Empty)
        End If
        context.Response.Clear()
        context.Response.StatusCode = 301
        context.Response.AppendHeader("location", url)
        context.Response.[End]()
    End Sub

    Public Sub Dispose() Implements IHttpModule.Dispose
    End Sub
End Class
